vortex-expr/src/pruning/pruning_predicate.rs

```diff
@@ -0,0 +1,81 @@
+use std::fmt::Display;
+
+use vortex_array::aliases::hash_map::HashMap;
+use vortex_array::aliases::hash_set::HashSet;
+use vortex_array::stats::Stat;
+use vortex_array::{Array, ArrayRef};
+use vortex_error::{VortexExpect as _, VortexResult};
+
+use super::field_or_identity::FieldOrIdentity;
+use super::pruning_predicate_rewriter::convert_to_pruning_expression;
+use super::relation::Relation;
+use crate::{ExprRef, Literal, Scope};
+
+#[derive(Debug, Clone)]
+pub struct PruningPredicate {
+    expr: ExprRef,
+    required_stats: Relation<FieldOrIdentity, Stat>,
+}
+
+impl Display for PruningPredicate {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "PruningPredicate({}, {{{}}})",
+            self.expr, self.required_stats
+        )
+    }
+}
+
+impl PruningPredicate {
+    pub fn try_new(original_expr: &ExprRef) -> Option<Self> {
+        let (expr, required_stats) = convert_to_pruning_expression(original_expr);
+        if let Some(lexp) = expr.as_any().downcast_ref::<Literal>() {
+            // Is the expression constant false, i.e. prune nothing
+            if lexp.value().as_bool_opt().and_then(|b| b.value()) == Some(false) {
+                return None;
+            }
+        }
+
+        Some(Self {
+            expr,
+            required_stats,
+        })
+    }
+
+    pub fn expr(&self) -> &ExprRef {
+        &self.expr
+    }
+
+    pub fn required_stats(&self) -> &HashMap<FieldOrIdentity, HashSet<Stat>> {
+        self.required_stats.map()
+    }
+
+    /// Evaluate this predicate against a per-chunk statistics table.
+    ///
+    /// Returns Ok(None) if any of the required statistics are not present in metadata.
+    /// If it returns Ok(Some(array)), the array is a boolean array with the same length as the
+    /// metadata, and a true value means the chunk _can_ be pruned.
+    pub fn evaluate(&self, metadata: &dyn Array) -> VortexResult<Option<ArrayRef>> {
```

Comment:
this is not relevant for this PR, but i'm wondering if the control flow is inverted.

I kind of thing that a thing that is prunable should receive a PruningPredicate, and then do something with it, rather than forcing the PruningPredicate to receive this type-erased thing that we pretty-promise looks like a stats table defined in the other crate

Reply:
Agreed.